File: serenity-app/src/main/java/us/nineworlds/serenity/injection/modules/providers/DataSourceFactoryProvider.java
package us.nineworlds.serenity.injection.modules.providers;

import android.content.Context;

import androidx.core.app.unusedapprestrictions.IUnusedAppRestrictionsBackportService;

import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.upstream.HttpDataSource;
import javax.inject.Inject;
import javax.inject.Provider;
import us.nineworlds.serenity.common.android.injection.ApplicationContext;

public class DataSourceFactoryProvider implements Provider<DataSource.Factory> {

  @Inject @ApplicationContext Context context;
  DefaultBandwidthMeter bandwidthMeter;
  @Inject HttpDataSource.Factory httpDataSourceFactory;

  @Override public DataSource.Factory get() {
    bandwidthMeter = DefaultBandwidthMeter.getSingletonInstance(context);

    return new DefaultDataSourceFactory(context, bandwidthMeter, httpDataSourceFactory);
  }
}

File: serenity-app/src/main/java/us/nineworlds/serenity/injection/modules/providers/DataSourceFactoryProvider.kt
package us.nineworlds.serenity.injection.modules.providers

import android.content.Context
import androidx.annotation.OptIn
import androidx.media3.common.util.UnstableApi
import androidx.media3.datasource.DataSource
import androidx.media3.datasource.DefaultDataSourceFactory
import androidx.media3.datasource.HttpDataSource
import androidx.media3.datasource.cache.CacheDataSource
import androidx.media3.exoplayer.upstream.DefaultBandwidthMeter
import javax.inject.Inject
import javax.inject.Provider
import us.nineworlds.serenity.SerenityApplication
import us.nineworlds.serenity.common.android.injection.ApplicationContext

class DataSourceFactoryProvider : Provider<DataSource.Factory> {
    @Inject
    @field:ApplicationContext
    lateinit var context: Context

    @Inject
    lateinit var httpDataSourceFactory: HttpDataSource.Factory

    @OptIn(UnstableApi::class)
    override fun get(): DataSource.Factory {
        val bandwidthMeter = DefaultBandwidthMeter.getSingletonInstance(context)
        val defaultDataSourceFactory =
            DefaultDataSourceFactory(context, bandwidthMeter, httpDataSourceFactory)
        return CacheDataSource.Factory()
            .setCache(SerenityApplication.simpleCache)
            .setUpstreamDataSourceFactory(defaultDataSourceFactory)

    }
}
